### ISSUE TITLE
Support GitHub API for content editing

### ISSUE THREAD

**Issue by SimonBiggs** [tinacms/tinacms#488]
## Summary

Allow the live generated site to be edited in it's production state by having users log in to GitHub and use TinaCMS to submit commits to GitHub via its API.

### Motivation

To support sites being able to be hosted on GitHub pages without any need to be backed by a server.

**Comment by ncphillips** [comment]
Thanks for the suggestion @SimonBiggs!

We did some explorations of this early on, but haven't gone very far in that direction yet. The main reason is that is this approach is extremely limited, to the point that it's practically impossible to support it for Gatsby sites. 

Why? It's very common to transform data as it's being added to Gatsby, and for GraphQL queries to modify that data (e.g. date formatting, images, etc). In order for Tina to work, it must have access to the original data, and that data must be able to be re-processed after it's updated. The more data processing that happens serverside, the less viable a serveless approach might be. Take the `gatsby-transformer-remark` for example, it does all of the processing to turn markdown into html serverside during the build step. The only way to allow for Markdown changes to be previewed would be to duplicate that process in the browser. That's just Gatsby though! If you setup an application that did all the processing client-side this approach could totally work. 

TL;DR – This is totally possible, but taking this approach _really_ limits how you can build your site.

**Comment by SimonBiggs** [comment]
That makes sense. I wonder if GitHub actions could be used to do this processing?

**Comment by SimonBiggs** [comment]
> That said, if you setup a site that did all the processing client-side this approach could totally work.

I might just give this a shot. My use case is I am building our church website:

https://github.com/Wagga-Church-of-Christ/wagga-church-of-christ
https://waggachurchofchrist.org.au/

It uses gatsby, free tier of netlify and the free tier of contentful at the moment. My requirements are that the ongoing costs are nil so that there isn't any extra load on the church budget, but also have it be easy to edit.

The website doesn't need to be fancy. Really just needs the basics. Might be possible to have all the processing be client side during edits, but still have the website be built on netlify so that users get a pre-built site.

What are your thoughts? Do you think it is worth me giving it a shot?

**Comment by ncphillips** [comment]
Absolutely! It would be amazing to see a setup like this for a simple site. If you need some guidance on how to do that (which isn’t exactly related to this issue) then start a  new topic in our [community forum](https://community.tinacms.org). I’m going to be afk for a couple days, but I can still reply occasionally

**Comment by dwalkr** [comment]
When it comes to processing markdown (or any content transformation) client-side, I think the best bet would be to duplicate the functionality of server-side processing. This allows you to have immediate feedback while editing, but still makes it possible to server-side render the HTML. SSR is beneficial for things like performance and SEO, and is a key feature of something like Gatsby, so it would be a regression to handle markdown _exclusively_ client-side.

@SimonBiggs I think this is what you mean, but wanted to clarify that this probably won't do what you want without implementing the processing in both the browser context and SSR context.

**Comment by SimonBiggs** [comment]
Yup, makes sense. Yup that was what I meant, but perhaps I had not
recognised the direct importance of it, and may have had my MVP ignore it.
Won't ignore it in the first iteration though now.

**Comment by ncphillips** [comment]
@SimonBiggs I made a demo of a create-react-app based site that uses Firebase as the backend. This demo does not need a backend server to be running. Check it out for [inspiration](https://github.com/ncphillips/demo-tinacms-firebase)

**Comment by SimonBiggs** [comment]
> @SimonBiggs I made a demo of a create-react-app based site that uses Firebase as the backend. This demo does not need a backend server to be running. Check it out for [inspiration](https://github.com/ncphillips/demo-tinacms-firebase)

Thank you! That will be a great starting point I believe.

**Comment by TomerAberbach** [comment]
I think [Netlify CMS Custom Previews](https://www.netlifycms.org/docs/customization) achieve similar functionality to what this issue is about. Might be worth your time to see how Netlify CMS does it since it uses the GitHub API.

The [Gatsby Netlify CMS starter](https://github.com/netlify-templates/gatsby-starter-netlify-cms/tree/master/src/cms/preview-templates) also serves as a good usage example.

**Comment by jpatters** [comment]
@TomerAberbach Netlify CMS requires that you create a separate template that renders the preview.

**Comment by kendallstrautman** [comment]
@SimonBiggs — we made a GitHub Client for the [open authoring workflow](https://tinacms.org/guides/nextjs/github-open-authoring/initial-setup). Right now it is coupled with Next.js, but it may be of interest!

**Comment by kendallstrautman** [comment]
At this time, you can edit content via the GitHub API with `react-tinacms-github` and `next-tinacms-github`. Closing this for now.
